src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryAccessor.cs

```diff
@@ -418,11 +420,16 @@ public Decimal ReadDecimal(Int64 position)
                     int hi = UnsafeReadInt32(pointer + 8);
                     int flags = UnsafeReadInt32(pointer + 12);
 
-                    int* ptr = (int*)(&d);
-                    *ptr++ = flags;
-                    *ptr++ = hi;
-                    *ptr++ = lo;
-                    *ptr = mid;
+                    // Check for invalid Decimal values
+                    if (!((flags & ~(SignMask | ScaleMask)) == 0 && (flags & ScaleMask) <= (28 << 16)))
+                    {
+                        throw new IOException(SR.IO_BadDecimal);
```

Comment:
This is a different exception type than is thrown from the mscorlib version, right?  If the goal of this change is to bring the behaviors in sync, seems we should throw the same exception type.